File: specs/typespec/org/company-addresses.go
package org

import (
	"fmt"

	"vetchium-api-server.typespec/common"
)

const (
	addressTitleMax        = 100
	addressLine1Max        = 200
	addressLine2Max        = 200
	addressCityMax         = 100
	addressStateMax        = 100
	addressPostalCodeMax   = 20
	addressCountryMax      = 100
	addressMapUrlMax       = 500
	addressMapUrlsMaxCount = 5

	errAddressTitleRequired     = "title is required"
	errAddressTitleTooLong      = "title must be at most 100 characters"
	errAddressLine1Required     = "address_line1 is required"
	errAddressLine1TooLong      = "address_line1 must be at most 200 characters"
	errAddressLine2TooLong      = "address_line2 must be at most 200 characters"
	errAddressCityRequired      = "city is required"
	errAddressCityTooLong       = "city must be at most 100 characters"
	errAddressStateTooLong      = "state must be at most 100 characters"
	errAddressPostalCodeTooLong = "postal_code must be at most 20 characters"
	errAddressCountryRequired   = "country is required"
	errAddressCountryTooLong    = "country must be at most 100 characters"
	errAddressMapUrlsTooMany    = "map_urls must have at most 5 entries"
	errAddressMapUrlTooLong     = "each map_url must be at most 500 characters"
	errAddressIDRequired        = "address_id is required"
	errAddressStatusInvalid     = "filter_status must be 'active' or 'disabled'"
)

type OrgAddressStatus string

const (
	OrgAddressStatusActive   OrgAddressStatus = "active"
	OrgAddressStatusDisabled OrgAddressStatus = "disabled"
)

type OrgAddress struct {
	AddressID    string           `json:"address_id"`
	Title        string           `json:"title"`
	AddressLine1 string           `json:"address_line1"`
	AddressLine2 *string          `json:"address_line2,omitempty"`
	City         string           `json:"city"`
	State        *string          `json:"state,omitempty"`
	PostalCode   *string          `json:"postal_code,omitempty"`
	Country      string           `json:"country"`
	MapUrls      []string         `json:"map_urls"`
	Status       OrgAddressStatus `json:"status"`
	CreatedAt    string           `json:"created_at"`
}

type CreateAddressRequest struct {
	Title        string   `json:"title"`
	AddressLine1 string   `json:"address_line1"`
	AddressLine2 *string  `json:"address_line2,omitempty"`
	City         string   `json:"city"`
	State        *string  `json:"state,omitempty"`
	PostalCode   *string  `json:"postal_code,omitempty"`
	Country      string   `json:"country"`
	MapUrls      []string `json:"map_urls,omitempty"`
}

func (r CreateAddressRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.Title == "" {
		errs = append(errs, common.NewValidationError("title", fmt.Errorf(errAddressTitleRequired)))
	} else if len(r.Title) > addressTitleMax {
		errs = append(errs, common.NewValidationError("title", fmt.Errorf(errAddressTitleTooLong)))
	}
	if r.AddressLine1 == "" {
		errs = append(errs, common.NewValidationError("address_line1", fmt.Errorf(errAddressLine1Required)))
	} else if len(r.AddressLine1) > addressLine1Max {
		errs = append(errs, common.NewValidationError("address_line1", fmt.Errorf(errAddressLine1TooLong)))
	}
	if r.AddressLine2 != nil && len(*r.AddressLine2) > addressLine2Max {
		errs = append(errs, common.NewValidationError("address_line2", fmt.Errorf(errAddressLine2TooLong)))
	}
	if r.City == "" {
		errs = append(errs, common.NewValidationError("city", fmt.Errorf(errAddressCityRequired)))
	} else if len(r.City) > addressCityMax {
		errs = append(errs, common.NewValidationError("city", fmt.Errorf(errAddressCityTooLong)))
	}
	if r.State != nil && len(*r.State) > addressStateMax {
		errs = append(errs, common.NewValidationError("state", fmt.Errorf(errAddressStateTooLong)))
	}
	if r.PostalCode != nil && len(*r.PostalCode) > addressPostalCodeMax {
		errs = append(errs, common.NewValidationError("postal_code", fmt.Errorf(errAddressPostalCodeTooLong)))
	}
	if r.Country == "" {
		errs = append(errs, common.NewValidationError("country", fmt.Errorf(errAddressCountryRequired)))
	} else if len(r.Country) > addressCountryMax {
		errs = append(errs, common.NewValidationError("country", fmt.Errorf(errAddressCountryTooLong)))
	}
	if len(r.MapUrls) > addressMapUrlsMaxCount {
		errs = append(errs, common.NewValidationError("map_urls", fmt.Errorf(errAddressMapUrlsTooMany)))
	} else {
		for _, u := range r.MapUrls {
			if len(u) > addressMapUrlMax {
				errs = append(errs, common.NewValidationError("map_urls", fmt.Errorf(errAddressMapUrlTooLong)))
				break
			}
		}
	}
	return errs
}

type UpdateAddressRequest struct {
	AddressID    string   `json:"address_id"`
	Title        string   `json:"title"`
	AddressLine1 string   `json:"address_line1"`
	AddressLine2 *string  `json:"address_line2,omitempty"`
	City         string   `json:"city"`
	State        *string  `json:"state,omitempty"`
	PostalCode   *string  `json:"postal_code,omitempty"`
	Country      string   `json:"country"`
	MapUrls      []string `json:"map_urls,omitempty"`
}

func (r UpdateAddressRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.AddressID == "" {
		errs = append(errs, common.NewValidationError("address_id", fmt.Errorf(errAddressIDRequired)))
	}
	if r.Title == "" {
		errs = append(errs, common.NewValidationError("title", fmt.Errorf(errAddressTitleRequired)))
	} else if len(r.Title) > addressTitleMax {
		errs = append(errs, common.NewValidationError("title", fmt.Errorf(errAddressTitleTooLong)))
	}
	if r.AddressLine1 == "" {
		errs = append(errs, common.NewValidationError("address_line1", fmt.Errorf(errAddressLine1Required)))
	} else if len(r.AddressLine1) > addressLine1Max {
		errs = append(errs, common.NewValidationError("address_line1", fmt.Errorf(errAddressLine1TooLong)))
	}
	if r.AddressLine2 != nil && len(*r.AddressLine2) > addressLine2Max {
		errs = append(errs, common.NewValidationError("address_line2", fmt.Errorf(errAddressLine2TooLong)))
	}
	if r.City == "" {
		errs = append(errs, common.NewValidationError("city", fmt.Errorf(errAddressCityRequired)))
	} else if len(r.City) > addressCityMax {
		errs = append(errs, common.NewValidationError("city", fmt.Errorf(errAddressCityTooLong)))
	}
	if r.State != nil && len(*r.State) > addressStateMax {
		errs = append(errs, common.NewValidationError("state", fmt.Errorf(errAddressStateTooLong)))
	}
	if r.PostalCode != nil && len(*r.PostalCode) > addressPostalCodeMax {
		errs = append(errs, common.NewValidationError("postal_code", fmt.Errorf(errAddressPostalCodeTooLong)))
	}
	if r.Country == "" {
		errs = append(errs, common.NewValidationError("country", fmt.Errorf(errAddressCountryRequired)))
	} else if len(r.Country) > addressCountryMax {
		errs = append(errs, common.NewValidationError("country", fmt.Errorf(errAddressCountryTooLong)))
	}
	if len(r.MapUrls) > addressMapUrlsMaxCount {
		errs = append(errs, common.NewValidationError("map_urls", fmt.Errorf(errAddressMapUrlsTooMany)))
	} else {
		for _, u := range r.MapUrls {
			if len(u) > addressMapUrlMax {
				errs = append(errs, common.NewValidationError("map_urls", fmt.Errorf(errAddressMapUrlTooLong)))
				break
			}
		}
	}
	return errs
}

type DisableAddressRequest struct {
	AddressID string `json:"address_id"`
}

func (r DisableAddressRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.AddressID == "" {
		errs = append(errs, common.NewValidationError("address_id", fmt.Errorf(errAddressIDRequired)))
	}
	return errs
}

type EnableAddressRequest struct {
	AddressID string `json:"address_id"`
}

func (r EnableAddressRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.AddressID == "" {
		errs = append(errs, common.NewValidationError("address_id", fmt.Errorf(errAddressIDRequired)))
	}
	return errs
}

type GetAddressRequest struct {
	AddressID string `json:"address_id"`
}

func (r GetAddressRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.AddressID == "" {
		errs = append(errs, common.NewValidationError("address_id", fmt.Errorf(errAddressIDRequired)))
	}
	return errs
}

type ListAddressesRequest struct {
	FilterStatus  *OrgAddressStatus `json:"filter_status,omitempty"`
	PaginationKey *string           `json:"pagination_key,omitempty"`
	Limit         *int32            `json:"limit,omitempty"`
}

func (r ListAddressesRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.FilterStatus != nil && *r.FilterStatus != OrgAddressStatusActive && *r.FilterStatus != OrgAddressStatusDisabled {
		errs = append(errs, common.NewValidationError("filter_status", fmt.Errorf(errAddressStatusInvalid)))
	}
	return errs
}

type ListAddressesResponse struct {
	Addresses         []OrgAddress `json:"addresses"`
	NextPaginationKey string       `json:"next_pagination_key,omitempty"`
}

File: api-schema/org/company-addresses.ts
import { type ValidationError, newValidationError } from "../common/common";

const ADDRESS_TITLE_MAX = 100;
const ADDRESS_LINE1_MAX = 200;
const ADDRESS_LINE2_MAX = 200;
const ADDRESS_CITY_MAX = 100;
const ADDRESS_STATE_MAX = 100;
const ADDRESS_POSTAL_CODE_MAX = 20;
const ADDRESS_COUNTRY_MAX = 100;
const ADDRESS_MAP_URL_MAX = 500;
const ADDRESS_MAP_URLS_MAX_ENTRIES = 5;

export const ERR_ADDRESS_TITLE_REQUIRED = "title is required";
export const ERR_ADDRESS_TITLE_TOO_LONG =
	"title must be at most 100 characters";
export const ERR_ADDRESS_LINE1_REQUIRED = "address_line1 is required";
export const ERR_ADDRESS_LINE1_TOO_LONG =
	"address_line1 must be at most 200 characters";
export const ERR_ADDRESS_LINE2_TOO_LONG =
	"address_line2 must be at most 200 characters";
export const ERR_ADDRESS_CITY_REQUIRED = "city is required";
export const ERR_ADDRESS_CITY_TOO_LONG = "city must be at most 100 characters";
export const ERR_ADDRESS_STATE_TOO_LONG =
	"state must be at most 100 characters";
export const ERR_ADDRESS_POSTAL_CODE_TOO_LONG =
	"postal_code must be at most 20 characters";
export const ERR_ADDRESS_COUNTRY_REQUIRED = "country is required";
export const ERR_ADDRESS_COUNTRY_TOO_LONG =
	"country must be at most 100 characters";
export const ERR_ADDRESS_MAP_URLS_TOO_MANY =
	"map_urls must have at most 5 entries";
export const ERR_ADDRESS_MAP_URL_TOO_LONG =
	"each map_url must be at most 500 characters";
export const ERR_ADDRESS_ID_REQUIRED = "address_id is required";
export const ERR_ADDRESS_STATUS_INVALID =
	"filter_status must be 'active' or 'disabled'";

export type OrgAddressStatus = "active" | "disabled";
export const OrgAddressStatusActive: OrgAddressStatus = "active";
export const OrgAddressStatusDisabled: OrgAddressStatus = "disabled";

export interface OrgAddress {
	address_id: string;
	title: string;
	address_line1: string;
	address_line2?: string;
	city: string;
	state?: string;
	postal_code?: string;
	country: string;
	map_urls: string[];
	status: OrgAddressStatus;
	created_at: string;
}

export interface CreateAddressRequest {
	title: string;
	address_line1: string;
	address_line2?: string;
	city: string;
	state?: string;
	postal_code?: string;
	country: string;
	map_urls?: string[];
}

export function validateCreateAddressRequest(
	r: CreateAddressRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!r.title)
		errs.push(newValidationError("title", ERR_ADDRESS_TITLE_REQUIRED));
	else if (r.title.length > ADDRESS_TITLE_MAX)
		errs.push(newValidationError("title", ERR_ADDRESS_TITLE_TOO_LONG));
	if (!r.address_line1)
		errs.push(newValidationError("address_line1", ERR_ADDRESS_LINE1_REQUIRED));
	else if (r.address_line1.length > ADDRESS_LINE1_MAX)
		errs.push(newValidationError("address_line1", ERR_ADDRESS_LINE1_TOO_LONG));
	if (
		r.address_line2 !== undefined &&
		r.address_line2.length > ADDRESS_LINE2_MAX
	)
		errs.push(newValidationError("address_line2", ERR_ADDRESS_LINE2_TOO_LONG));
	if (!r.city) errs.push(newValidationError("city", ERR_ADDRESS_CITY_REQUIRED));
	else if (r.city.length > ADDRESS_CITY_MAX)
		errs.push(newValidationError("city", ERR_ADDRESS_CITY_TOO_LONG));
	if (r.state !== undefined && r.state.length > ADDRESS_STATE_MAX)
		errs.push(newValidationError("state", ERR_ADDRESS_STATE_TOO_LONG));
	if (
		r.postal_code !== undefined &&
		r.postal_code.length > ADDRESS_POSTAL_CODE_MAX
	)
		errs.push(
			newValidationError("postal_code", ERR_ADDRESS_POSTAL_CODE_TOO_LONG)
		);
	if (!r.country)
		errs.push(newValidationError("country", ERR_ADDRESS_COUNTRY_REQUIRED));
	else if (r.country.length > ADDRESS_COUNTRY_MAX)
		errs.push(newValidationError("country", ERR_ADDRESS_COUNTRY_TOO_LONG));
	if (r.map_urls !== undefined) {
		if (r.map_urls.length > ADDRESS_MAP_URLS_MAX_ENTRIES)
			errs.push(newValidationError("map_urls", ERR_ADDRESS_MAP_URLS_TOO_MANY));
		else
			for (const url of r.map_urls) {
				if (url.length > ADDRESS_MAP_URL_MAX) {
					errs.push(
						newValidationError("map_urls", ERR_ADDRESS_MAP_URL_TOO_LONG)
					);
					break;
				}
			}
	}
	return errs;
}

export interface UpdateAddressRequest {
	address_id: string;
	title: string;
	address_line1: string;
	address_line2?: string;
	city: string;
	state?: string;
	postal_code?: string;
	country: string;
	map_urls?: string[];
}

export function validateUpdateAddressRequest(
	r: UpdateAddressRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!r.address_id)
		errs.push(newValidationError("address_id", ERR_ADDRESS_ID_REQUIRED));
	if (!r.title)
		errs.push(newValidationError("title", ERR_ADDRESS_TITLE_REQUIRED));
	else if (r.title.length > ADDRESS_TITLE_MAX)
		errs.push(newValidationError("title", ERR_ADDRESS_TITLE_TOO_LONG));
	if (!r.address_line1)
		errs.push(newValidationError("address_line1", ERR_ADDRESS_LINE1_REQUIRED));
	else if (r.address_line1.length > ADDRESS_LINE1_MAX)
		errs.push(newValidationError("address_line1", ERR_ADDRESS_LINE1_TOO_LONG));
	if (
		r.address_line2 !== undefined &&
		r.address_line2.length > ADDRESS_LINE2_MAX
	)
		errs.push(newValidationError("address_line2", ERR_ADDRESS_LINE2_TOO_LONG));
	if (!r.city) errs.push(newValidationError("city", ERR_ADDRESS_CITY_REQUIRED));
	else if (r.city.length > ADDRESS_CITY_MAX)
		errs.push(newValidationError("city", ERR_ADDRESS_CITY_TOO_LONG));
	if (r.state !== undefined && r.state.length > ADDRESS_STATE_MAX)
		errs.push(newValidationError("state", ERR_ADDRESS_STATE_TOO_LONG));
	if (
		r.postal_code !== undefined &&
		r.postal_code.length > ADDRESS_POSTAL_CODE_MAX
	)
		errs.push(
			newValidationError("postal_code", ERR_ADDRESS_POSTAL_CODE_TOO_LONG)
		);
	if (!r.country)
		errs.push(newValidationError("country", ERR_ADDRESS_COUNTRY_REQUIRED));
	else if (r.country.length > ADDRESS_COUNTRY_MAX)
		errs.push(newValidationError("country", ERR_ADDRESS_COUNTRY_TOO_LONG));
	if (r.map_urls !== undefined) {
		if (r.map_urls.length > ADDRESS_MAP_URLS_MAX_ENTRIES)
			errs.push(newValidationError("map_urls", ERR_ADDRESS_MAP_URLS_TOO_MANY));
		else
			for (const url of r.map_urls) {
				if (url.length > ADDRESS_MAP_URL_MAX) {
					errs.push(
						newValidationError("map_urls", ERR_ADDRESS_MAP_URL_TOO_LONG)
					);
					break;
				}
			}
	}
	return errs;
}

export interface DisableAddressRequest {
	address_id: string;
}
export function validateDisableAddressRequest(
	r: DisableAddressRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!r.address_id)
		errs.push(newValidationError("address_id", ERR_ADDRESS_ID_REQUIRED));
	return errs;
}

export interface EnableAddressRequest {
	address_id: string;
}
export function validateEnableAddressRequest(
	r: EnableAddressRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!r.address_id)
		errs.push(newValidationError("address_id", ERR_ADDRESS_ID_REQUIRED));
	return errs;
}

export interface GetAddressRequest {
	address_id: string;
}
export function validateGetAddressRequest(
	r: GetAddressRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (!r.address_id)
		errs.push(newValidationError("address_id", ERR_ADDRESS_ID_REQUIRED));
	return errs;
}

export interface ListAddressesRequest {
	filter_status?: OrgAddressStatus;
	pagination_key?: string;
	limit?: number;
}

export function validateListAddressesRequest(
	r: ListAddressesRequest
): ValidationError[] {
	const errs: ValidationError[] = [];
	if (
		r.filter_status !== undefined &&
		r.filter_status !== OrgAddressStatusActive &&
		r.filter_status !== OrgAddressStatusDisabled
	) {
		errs.push(newValidationError("filter_status", ERR_ADDRESS_STATUS_INVALID));
	}
	return errs;
}

export interface ListAddressesResponse {
	addresses: OrgAddress[];
	next_pagination_key?: string;
}
